Upload-only link shares

PVince81 opened the pull request:
Fixes https://github.com/owncloud/core/issues/7747.

Checklist here: https://github.com/owncloud/core/issues/7747#issuecomment-282263457

There's more to come, we can't merge this as is because backend support is required.

Commit:
Add checkbox to remove listing permission in public links

Commit:
Let the upload-only permission pass through OCS Share API

Commit:
Allow setting create-only permission in share manager

Commit:
Upload-only form for public link page

PVince81 commented:
- [ ] BUG: how to solve conflicts ?

- [ ] TODO: JS unit tests for the new view
- [ ] TODO: beautify the view (cc @felixheidecke)

PVince81 commented:
This is using the file-upload.js code so when this https://github.com/owncloud/core/pull/26306#issuecomment-290762884 is fixed to work with chunking, the button from anonymous upload should also be able to work with chunking.

@DeepDiver1975 @tomneedham @IljaN @butonic check this out

PVince81 commented:
- [ ] TODO: error handling, properly display error message
   - [ ] conflict
   - [ ] folder gone / whatever
   - [ ] insufficient space

PVince81 commented:
@pmaier1 I think having it called "Allow file listing" and set by default is better.

This is how it looks on this PR:

![files_drop_core](https://cloud.githubusercontent.com/assets/277525/25657104/6d122bea-2ffc-11e7-8227-b7f88e8a2b4a.png)

If you disable "Allow uploads" the option "Allow file listing" becomes greyed out and checked.

pmaier1 commented:
Ok nice so far. I don't like the wording "Allow viewing file listing" :) "Hide file listing" would be perfect IMO.
As you would need to change your logic probably call it "Show file listing".

PVince81 commented:
The part I dislike about "Hide file listing" is mostly because you need to think positive for "Allow uploads" and then negative "Hide" instead of "Show".

We could also rename "Allow uploads" to "Disallow uploads".

@felixheidecke any feedback on the wordings ?

pmaier1 commented:
I leave this up to you. 'Allow' and 'Show' or 'Disallow' and 'Hide' is both fine for me while I like the first pair a bit more :)

PVince81 commented:
- [ ] TODO: add capability entry

PVince81 commented:
- [ ] Change "Allow upload" to "Allow editing"

PVince81 commented:
- [ ] "Show file listing"

Commit:
Add integration tests for upload-only feature

PVince81 commented:
Added integration tests.

One covers the file name deduplication and will fail currently as I still need to implement this.

Commit:
Adjust dialog wordings

Commit:
Add capability for upload-only feature

PVince81 commented:
Capability added "supports_upload_only"

Commit:
Prevent download non-readable link share

Commit:
Add Sabre Autorename plugin for public upload

Introduces a new header X-OC-Autorename that tells the autorename plugin
to automatically rename a file if the target already exists.

Currently only used and applied for upload-only shares.

PVince81 commented:
Found a relatively clean way to deduplicate names by overriding the PUT method, but only when a header "X-OC-Autorename" is set.

@DeepDiver1975 can you have a look and tell me whether the approach is acceptable ? It's all in https://github.com/owncloud/core/pull/27548/commits/c47b09b2987a0fdde9a5dacea492cbb9a22b6b63

PVince81 commented:
As discussed with @DeepDiver1975 in a chat, the approach is acceptable.
One change:

- [ ] rename "X-OC-Autorename" to "OC-Autorename"

Commit:
Remove confused and unused port arg in OC.Files.Client constructor

Commit:
Add JS unit tests for PublicUploadView

PVince81 commented:
Added JS unit tests.

Next up: beautify and proper error messages

tomneedham commented:
@butonic

PVince81 commented:
Uh oh, seems the push of JS unit tests from the other computer didn't work. Note to self: no rebase!

Commit:
Throw NotFound exception when link share source not found

Plus some bonus integration tests

Commit:
Improve error handling for public upload, generic messages though

PVince81 commented:
- [ ] cherry pick commits from my other computer/branch which weren't pushed properly (OC-Autorename + JS unit tests)

PVince81 commented:
- [ ] TEST: drag and drop upload

PVince81 commented:
- [ ] design adjustments:
    - [ ] visible dropzone with text "Select files for upload"
        - [ ] clicking dropzone opens file picker
        - [ ] dropping on dropzone uploads files directly

Commit:
Restyled public upload-only drop zone

Commit:
Added progress bar for public upload-only page

PVince81 commented:
![files-drop](https://cloud.githubusercontent.com/assets/277525/25970094/bc3075fa-3697-11e7-8f0c-296885d9acdd.gif)

SergioBertolinSG commented on tests/integration/features/sharing-v1.feature:
```diff
@@ -263,6 +264,27 @@ Feature: sharing
+      | path | FOLDER |
+      | shareType | 3 |
+      | permissions | 4 | 
+    And User "user0" deletes file "/FOLDER"
```
This should be 'When'.

SergioBertolinSG commented on tests/integration/features/sharing-v1.feature:
```diff
@@ -263,6 +264,27 @@ Feature: sharing
+      | shareType | 3 |
+      | permissions | 4 | 
+    And User "user0" deletes file "/FOLDER"
+	Then publicly uploading a file does not work
```
Too much indentation here.

SergioBertolinSG commented on tests/integration/features/sharing-v1.feature:
```diff
@@ -216,6 +216,75 @@ Feature: sharing
+      | shareType | 3 |
+      | password | publicpw |
+      | permissions | 4 | 
+	And publicly uploading file "test.txt" with password "publicpw" and content "test"
```
Indentation from here to the end of the scenario is not aligned.

SergioBertolinSG commented on tests/integration/features/sharing-v1.feature:
```diff
@@ -216,6 +216,75 @@ Feature: sharing
+      | path | FOLDER |
+      | shareType | 3 |
+      | permissions | 4 | 
+	And publicly uploading file "test.txt" with content "test"
```
Indentation from here to the end of the scenario is not aligned.

SergioBertolinSG commented on tests/integration/features/sharing-v1.feature:
```diff
@@ -216,6 +216,75 @@ Feature: sharing
+      | path | FOLDER |
+      | shareType | 3 |
+      | permissions | 4 | 
+	And publicly uploading file "test.txt" with content "test"
```
Indentation from here to the end of the scenario is not aligned.

Commit:
Tabify behat file for sharing tests

Commit:
Add missing Whens

SergioBertolinSG commented on tests/integration/features/sharing-v1.feature:
```diff
@@ -1,1162 +1,1231 @@
+			| file_parent | A_NUMBER |
+			| share_with_displayname | user1 |
+			| displayname_owner | user0 |
+			| mimetype			| text/plain |
```
Too separated.

PVince81 replied:
I did a search and replace of two spaces "  " with tabs. Apparently for some reason there were two spaces there. I'll adjust it.

SergioBertolinSG commented on tests/integration/features/sharing-v1.feature:
```diff
@@ -1,1162 +1,1231 @@
+			| storage_id | home::user0 |
+			| file_parent | A_NUMBER |
+			| displayname_owner | user0 |
+			| mimetype			| text/plain |
```
Too separated.

PVince81 commented:
Rebased and fixes review comments

PVince81 commented:
Going to pre-review the diff and fix missing stuff.

PVince81 commented:
Ok, fixed everything I found when looking at the diff.

Please review! See whitespace-less diff: https://github.com/owncloud/core/pull/27548/files?w=1

@pmaier1 @DeepDiver1975 @tomneedham @felixheidecke

felixheidecke commented on apps/files_sharing/js/PublicUploadView.js:
```diff
@@ -0,0 +1,150 @@
+		render: function () {
+			this.$el.html(this.template({
+				title: t('files_sharing', 'Anonymous upload'),
+				uploadButtonLabel: t('files_sharing', 'Select files for upload'),
```
I suggest *"Drop files here for upload"* to hint to the drag & drop ability.

pmaier1 replied:
"Click here to select files or use drag & drop to upload" as you can also use the file picker I assume.

PVince81 replied:
I'll wait for you guys to settle on a solution, let me know 😄

pmaier1 replied:
My final call: "Click to select files or use drag & drop to upload"

DeepDiver1975 reviewed (approved):
NICE

PVince81 commented:
I'll go with @pmaier1's text for now and we can merge.

If it's not optimal we can change it in a subsequent PR. This will make it possible to already get this tested...

Commit:
Update upload-only drop area text

PVince81 commented:
Tests passed on previous commit and the text is not covered by unit tests => merge

PVince81 commented:
> Tests passed on previous commit and the text is not covered by unit tests => merge

Never trust the green ticks, Jenkins wasn't prevent, there's a failure on master now 💥

butonic commented:
Is there a migration path when upgrading from an oc 9.x with files_drop app?

PVince81 commented:
@butonic no migration path possible for technical reasons (route name is different).

It was decided to say in the release notes that the shares must be recreated.

cc @pmaier1

ckamm commented:
@PVince81 I'm adding the relevant functionality to the client. To clarify: Previously the client checked the "Allow editing" box if create and update permission were present. This seems to have changed? To get matching behavior I must only check for create permission.

Does the update permission still toggle anything ui wise?
